File: results/0.75/ELF/elf-deepseek-chat.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

// Define ELF header structure
typedef struct {
    uint8_t e_ident[16];
    uint16_t e_type;
    uint16_t e_machine;
    uint32_t e_version;
    uint64_t e_entry;
    uint64_t e_phoff;
    uint64_t e_shoff;
    uint32_t e_flags;
    uint16_t e_ehsize;
    uint16_t e_phentsize;
    uint16_t e_phnum;
    uint16_t e_shentsize;
    uint16_t e_shnum;
    uint16_t e_shstrndx;
} Elf64_Ehdr;

// Define ELF program header structure
typedef struct {
    uint32_t p_type;
    uint32_t p_flags;
    uint64_t p_offset;
    uint64_t p_vaddr;
    uint64_t p_paddr;
    uint64_t p_filesz;
    uint64_t p_memsz;
    uint64_t p_align;
} Elf64_Phdr;

// Define ELF section header structure
typedef struct {
    uint32_t sh_name;
    uint32_t sh_type;
    uint64_t sh_flags;
    uint64_t sh_addr;
    uint64_t sh_offset;
    uint64_t sh_size;
    uint32_t sh_link;
    uint32_t sh_info;
    uint64_t sh_addralign;
    uint64_t sh_entsize;
} Elf64_Shdr;

// Parser for ELF header
HParser *elf_header_parser() {
    return h_sequence(
        h_many(h_uint8()), // e_ident
        h_uint16(),        // e_type
        h_uint16(),        // e_machine
        h_uint32(),        // e_version
        h_uint64(),        // e_entry
        h_uint64(),        // e_phoff
        h_uint64(),        // e_shoff
        h_uint32(),        // e_flags
        h_uint16(),        // e_ehsize
        h_uint16(),        // e_phentsize
        h_uint16(),        // e_phnum
        h_uint16(),        // e_shentsize
        h_uint16(),        // e_shnum
        h_uint16(),        // e_shstrndx
        NULL
    );
}

// Parser for ELF program header
HParser *elf_program_header_parser() {
    return h_sequence(
        h_uint32(), // p_type
        h_uint32(), // p_flags
        h_uint64(), // p_offset
        h_uint64(), // p_vaddr
        h_uint64(), // p_paddr
        h_uint64(), // p_filesz
        h_uint64(), // p_memsz
        h_uint64(), // p_align
        NULL
    );
}

// Parser for ELF section header
HParser *elf_section_header_parser() {
    return h_sequence(
        h_uint32(), // sh_name
        h_uint32(), // sh_type
        h_uint64(), // sh_flags
        h_uint64(), // sh_addr
        h_uint64(), // sh_offset
        h_uint64(), // sh_size
        h_uint32(), // sh_link
        h_uint32(), // sh_info
        h_uint64(), // sh_addralign
        h_uint64(), // sh_entsize
        NULL
    );
}

// Main function
int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <binary file>\n", argv[0]);
        return 1;
    }

    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        perror("Failed to open file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    uint8_t *buffer = (uint8_t *)malloc(file_size);
    if (!buffer) {
        perror("Failed to allocate memory");
        fclose(file);
        return 1;
    }

    fread(buffer, 1, file_size, file);
    fclose(file);

    HParser *elf_parser = elf_header_parser();
    HParseResult *result = h_parse(elf_parser, buffer, file_size);

    if (!result) {
        fprintf(stderr, "Failed to parse ELF header\n");
        free(buffer);
        return 1;
    }

    // Access parsed data
    Elf64_Ehdr *elf_header = (Elf64_Ehdr *)result->ast;

    // Print some fields from the ELF header
    printf("ELF Type: %u\n", elf_header->e_type);
    printf("Entry Point: 0x%lx\n", elf_header->e_entry);

    // Free resources
    h_parse_result_free(result);
    free(buffer);

    return 0;
}

File: results/0.75/ELF/elf-deepseek-chat.rs
use nom::{
    bytes::complete::{tag, take},
    number::complete::{le_u16, le_u32, le_u64, u8},
    sequence::tuple,
    IResult,
};
use std::{
    env,
    fs::File,
    io::{self, Read},
};

#[derive(Debug)]
struct ElfHeader {
    magic: [u8; 4],
    class: u8,
    data: u8,
    version: u8,
    os_abi: u8,
    abi_version: u8,
    pad: [u8; 7],
    e_type: u16,
    e_machine: u16,
    e_version: u32,
    e_entry: u64,
    e_phoff: u64,
    e_shoff: u64,
    e_flags: u32,
    e_ehsize: u16,
    e_phentsize: u16,
    e_phnum: u16,
    e_shentsize: u16,
    e_shnum: u16,
    e_shstrndx: u16,
}

fn parse_elf_header(input: &[u8]) -> IResult<&[u8], ElfHeader> {
    let (input, (magic, class, data, version, os_abi, abi_version, pad, e_type, e_machine, e_version, e_entry, e_phoff, e_shoff, e_flags, e_ehsize, e_phentsize, e_phnum, e_shentsize, e_shnum, e_shstrndx)) = tuple((
        tag(b"\x7FELF"),
        u8,
        u8,
        u8,
        u8,
        u8,
        take(7usize),
        le_u16,
        le_u16,
        le_u32,
        le_u64,
        le_u64,
        le_u64,
        le_u32,
        le_u16,
        le_u16,
        le_u16,
        le_u16,
        le_u16,
        le_u16,
    ))(input)?;

    Ok((
        input,
        ElfHeader {
            magic: *magic,
            class,
            data,
            version,
            os_abi,
            abi_version,
            pad: *pad,
            e_type,
            e_machine,
            e_version,
            e_entry,
            e_phoff,
            e_shoff,
            e_flags,
            e_ehsize,
            e_phentsize,
            e_phnum,
            e_shentsize,
            e_shnum,
            e_shstrndx,
        },
    ))
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <elf_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_elf_header(&buffer) {
        Ok((_, header)) => println!("{:#?}", header),
        Err(e) => eprintln!("Failed to parse ELF header: {:?}", e),
    }

    Ok(())
}